### ISSUE TITLE
prompt disappears w/ readline 8.1

### ISSUE THREAD

**Issue by obfusk** [hanslub42/rlwrap#108]
## Bug Report

With `libreadline8=8.0`:

```
$ rlwrap -C test python3 -c 'while True: print("you said", input(">>> "))'
>>> foo
you said foo
>>> bar
you said bar
>>> ^D
```

With `libreadline8=8.1`:

```
$ rlwrap -C test python3 -c 'while True: print("you said", input(">>> "))'
foo
you said foo
bar
you said bar
>>> 
```

The `>>> ` prompt is initially shown each time, but then disappears from the output afterwards.  If I input `baz` next, I get:

```
foo
you said foo
bar
you said bar
baz
you said baz
>>>
```

## Bug Analysis

* readline 8.1 enables bracketed-paste by default;
* when bracketed-paste is enabled, `rl_deprep_terminal()` prints `BRACK_PASTE_FINI` (`"\033[?2004l\r"`) to `rl_outstream`;
* this moves the cursor to the beginning of the line (instead of the position after the prompt), which causes the `rlwrap` prompt to be overwritten during `rl_redisplay()`.
* thus, adding `set enable-bracketed-paste off` to my `~/.inputrc` fixes the problem;
* THB I'm not entirely sure whether this is a bug in `readline` or in `rlwrap`, so I'm reporting it to both.

**Comment by hanslub42** [comment]
Your analysis is correct, and this is a `rlwrap` bug that has been there for a long time, and could have been discovered earlier. Since version 8.1, readline enables bracketed-paste by default that bug is now exposed. The simple fix, for now, is to disable bracketed-paste from within `rlwrap`, but since it is a useful feature I'll try to accomodate it.

**Comment by hanslub42** [comment]
This should be fixed (for now) by 7e3d0a772b93d405698f7604b85fc7a8878dff01

**Comment by hanslub42** [comment]
Close for now (even though bracketed-paste should be accommodated eventually)

**Comment by linas** [comment]
Alas, Debian stable seems to have upgraded recently to `libreadline8 8.1-1`  but still runs `rlwrap 0.43` .  I'm confused, because this all seems to have worked last week!?  Anyone able to ping the Debian maintainer?

**Comment by obfusk** [comment]
> Debian stable seems to have upgraded recently to `libreadline8 8.1-1`

I'm fairly certain bullseye (released on 2021-08-14) shipped with `8.1-1`, which was uploaded to `unstable` on 2020-12-08 and migrated to `testing` on 2020-12-13.

**Comment by hanslub42** [comment]
The story so far:

- `readline-8.1` switched the `enable-bracketed-paste` setting from _off_ by default to _on_ by default. This exposed the  long-standing `rlwrap` bug mentioned above (see also #113)  This was fixed in `rlwrap-0.44` (by disabling bracketed paste) and then again in `rlwrap-0.45` (by correctly handling it)
- Then, `readline-8.2` changed the positioning of the cursor after accepting a line (so that the fixes for `rlwrap-0.44` actually caused the prompt to be right-shifted (#168)). This was fixed in rlwrap-0.46.1 (but will disable bracketed paste when linked with `readline-8.1`

In summary: 
- `readline-8.1` needs at least `rlwrap-0.44` (preferabley even  `0.45` for correctly handling bracketed paste). It will work with `rlwrap-0.46.1` - with a force-disabled `bracketed-paste` setting, however.
- `readline-8.2` needs at least `rlwrap-0.46.1` 

@linas: I think the best way to go about it is to  [file a bug](https://packages.debian.org/stable/utils/reportbug) with Debian, referring to this discussion.

**Comment by obfusk** [comment]
I can confirm everything works w/ `bracketed-paste` re-enabled in Debian *un*stable (`8.2-1.3` / `0.46.1-1`) now.
But yes, Debian stable (`8.1-1` / `0.43-1+b2`) should still have the bug.
It would have had it since its release in 2021 though; nothing changed recently.

**Comment by linas** [comment]
Thank you! I opened https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1029664 for Debian stable. I use `rlwrap` all the time, on multiple OS'es and only spotted this issue a few days ago. Maybe I just wasn't using it on Debian stable all that much.

**Comment by obfusk** [comment]
@linas did you see the workaround for older versions? I put this in my `~/.inputrc`:

```inputrc
$include /etc/inputrc
# for rlwrap
set enable-bracketed-paste off
```

**Comment by linas** [comment]
thx!
